### PR TITLE
[not ready] client side hillshading with terrain-rgb raster tiles

### PR DESCRIPTION
**Pull request by mollymerp**
still very very rough here. will update this ticket soon w the to-do list before this is ready for review. 

🗻
 

## Launch Checklist

 - [ ] briefly describe the changes in this PR
 - [ ] write tests for all new functionality
 - [ ] document any changes to public APIs
 - [ ] post benchmark scores
 - [ ] manually test the debug page

### CODE REVIEW

**Comment by andrewharvey** [comment]
Would it make more sense to call the layer type `hillshade`? I think we should try to use terminology which fits in with future 3D terrain support. I've proposed some ideas about terminology over at https://github.com/mapbox/mapbox-gl-js/issues/1489#issuecomment-287611953. Over there I've proposed using `terrain` as the source type for terrain-rgb tiles, and `elevation` as the layer type for 3D surface.

**Comment by mollymerp** [comment]
Here's a rough list of what needs to happen before this can be merged into master:
- [ ] move mip map level creation + resampling to a worker thread
- [ ] fix tile seams with better design for backfilling tile borders
- [ ] get blending to work with other layers
- [ ] clean up shader debug code
- [ ] reduce shader attribute counts for compatibility with native
- [ ] implement light to behave consistently with extrusions
- [ ] finalize layer+source properties in the style spec #4711

**Comment by kkaefer** [comment]
We could move the preparation step for terrain textures to the beginning of a frame to improve rendering performance.

Right now, the workflow of a frame is like this:
* (main framebuffer is bound)
* Clear main framebuffer (✅ first operation is clear -> no restore)
* Render some layers
* Render terrain layer
  * For non-preprocessed tiles
    * **Bind terrain framebuffer**
    * Clear terrain framebuffer (✅ first operation is clear -> no restore)
    * Draw prepare step
    * **Bind main framebuffer**
  * Draw preprocessed terrain tiles
* Draw other layers (:warning: FBO restore)

When binding a framebuffer and rendering operations on it, the GPU has to restore the contents of the framebuffer. The driver can optimize that when we do a clear as the first drawing operation (since that discards the contents of the framebuffer). Therefore, drawing to the main framebuffer, then binding a *different* framebuffer, and rebinding the main framebuffer triggers a framebuffer load on the first draw call.

Instead, we could do all operations that require binding a separate framebuffer (computing the preprocessed textures) as the first thing when rendering a frame (i.e. before calling clear on the main framebuffer) and have the following order:

* (main framebuffer is bound)
* Preprocess all non-preprocessed terrain tiles (if any)
  * **Bind terrain framebuffer**
  * Clear terrain framebuffer (✅ first operation is clear -> no restore)
  * Draw prepare step
  * **Bind main framebuffer**
* Clear main framebuffer (✅ first operation is clear -> no restore)
* Render some layers
* Render terrain layer
  * Draw preprocessed terrain tiles
* Draw other layers

**Comment by kkaefer** [comment]
We'd likely need something like https://github.com/mapbox/mapbox-gl-native/issues/9415 to prevent flickering from overlapping parent/child tiles.

File: src/source/image_source.js [hunk 1]
```diff
@@ -142,12 +142,11 @@ class ImageSource extends Evented {
...
         this.tiles[tile.coord.w] = tile;
-        const maxInt16 = 32767;
         const array = new RasterBoundsArray();
         array.emplaceBack(this._tileCoords[0].x, this._tileCoords[0].y, 0, 0);
-        array.emplaceBack(this._tileCoords[1].x, this._tileCoords[1].y, maxInt16, 0);
-        array.emplaceBack(this._tileCoords[3].x, this._tileCoords[3].y, 0, maxInt16);
-        array.emplaceBack(this._tileCoords[2].x, this._tileCoords[2].y, maxInt16, maxInt16);
+        array.emplaceBack(this._tileCoords[1].x, this._tileCoords[1].y, EXTENT, 0);
+        array.emplaceBack(this._tileCoords[3].x, this._tileCoords[3].y, 0, EXTENT);
+        array.emplaceBack(this._tileCoords[2].x, this._tileCoords[2].y, EXTENT, EXTENT);
```
Comment (mollymerp) [inline src/source/image_source.js]: yep -- these changes are in #4770 and https://github.com/mapbox/mapbox-gl-native/pull/9153, but they're blocked until text-pitch changes are merged into native

**Comment by mollymerp** [comment]
this work is now tracked at #5286

**Pull request closed without merging**